core/controllers/learner_dashboard.py

```diff
@@ -0,0 +1,206 @@
+                'category': collection_summary.category,
+                'objective': collection_summary.objective,
+                'language_code': collection_summary.language_code,
+                'last_updated': utils.get_time_in_millisecs(
+                    collection_summary.collection_model_last_updated),
+                'created_on': utils.get_time_in_millisecs(
+                    collection_summary.collection_model_created_on),
+                'status': collection_summary.status,
+                'node_count': collection_summary.node_count,
+                'community_owned': collection_summary.community_owned,
+                'thumbnail_icon_url': (
+                    utils.get_thumbnail_icon_url_for_category(
+                        collection_summary.category)),
+                'thumbnail_bg_color': utils.get_hex_color_for_category(
+                    collection_summary.category),
+            })
+
+        for collection_summary in incomplete_collection_summaries:
+            incomplete_collection_summary_dicts.append({
+                'id': collection_summary.id,
+                'title': collection_summary.title,
+                'category': collection_summary.category,
+                'objective': collection_summary.objective,
+                'language_code': collection_summary.language_code,
+                'last_updated': utils.get_time_in_millisecs(
+                    collection_summary.collection_model_last_updated),
+                'created_on': utils.get_time_in_millisecs(
+                    collection_summary.collection_model_created_on),
+                'status': collection_summary.status,
+                'node_count': collection_summary.node_count,
+                'community_owned': collection_summary.community_owned,
+                'thumbnail_icon_url': (
+                    utils.get_thumbnail_icon_url_for_category(
+                        collection_summary.category)),
+                'thumbnail_bg_color': utils.get_hex_color_for_category(
+                    collection_summary.category),
+            })
+
+        creators_subscribed_to = (
+            subscription_services.get_all_creators_subscribed_to(self.user_id))
+        creators_settings = user_services.get_users_settings(
+            creators_subscribed_to)
+        subscription_list = []
+
+        for index, creator_settings in enumerate(creators_settings):
+            subscription_summary = {
+                'creator_picture_data_url': (
+                    creator_settings.profile_picture_data_url),
+                'creator_username': creator_settings.username,
+                'creator_impact': (
+                    user_services.get_user_impact_score(
+                        creators_subscribed_to[index]))
+            }
+
+            subscription_list.append(subscription_summary)
+
+        self.values.update({
+            'completed_explorations_list': completed_exp_summary_dicts,
```

Comment:
Ah, sorry, no. What I forgot to point out was that ndb.get_multi() allows you to fetch multiple entities at a time even though they're of different types. See https://cloud.google.com/appengine/docs/standard/python/ndb/creating-entities#Python_Batch_operations

In your PR you are doing individual gets for each of the "get info about user progress" components, right? The point I was making was that perhaps these could be combined. So maybe you need just one (or two) datastore GET calls at the outset to get all the entities you need first, and then you can process all the domain objects as needed.